tools/pip-compile-wrapper.py

```diff
@@ -0,0 +1,300 @@
+        print(f"Installing {pip_tools_req}...", end="")
+        sys.stdout.flush()
+        process = subprocess.run(
+            [sys.executable, "-m", "pip", "install", pip_tools_req],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        process.check_returncode()
+        print("\b\b Done.")
+
+
+def compile_source(
+    upgrade_all: bool = False, verbose: bool = False, dry_run: bool = False
+) -> None:
+    for batch in _resolve_deps(SOURCE_PATHS.keys()):
+        for target in batch:
+            _run_pip_compile(target, upgrade_all=upgrade_all, verbose=verbose, dry_run=dry_run)
+
+
+def upgrade_source(
+    upgrade_package_names: Set[str], verbose: bool = False, dry_run: bool = False
+) -> None:
+    packages_to_sources = _get_sources_for_packages(upgrade_package_names, TargetType.ALL)
+
+    _newline = "\n  - "
+    missing_packages = upgrade_package_names - packages_to_sources.keys()
+    if missing_packages:
+        print(
+            "Some of the given packages were not found in either source or target files.\n"
+            "Please check that the packages are spelled correctly.\n"
+            "If any of these packages were newly added to any of the source files you need to "
+            f"run '{SCRIPT_NAME} compile' first.\n"
+            f"Missing package(s):\n  - {_newline.join(missing_packages)}"
+        )
+        sys.exit(1)
+    grouped_packages_to_sources = [
+        (set(package_name for package_name, _ in group), key)
+        for key, group in groupby(
+            sorted(packages_to_sources.items(), key=itemgetter(1)), key=itemgetter(1)
+        )
+    ]
+
+    for package_names, source_names in grouped_packages_to_sources:
+        print(f"Upgrading package(s):\n  - {_newline.join(package_names)}")
+        for batch in _resolve_deps(source_names):
+            for batch_source in batch:
+                _run_pip_compile(
+                    batch_source, upgrade_packages=package_names, verbose=verbose, dry_run=dry_run
+                )
+
+
+def main() -> None:
+    op = OptionParser(
+        f"Usage: {SCRIPT_NAME} [options] compile\n"
+        f"       {SCRIPT_NAME} [options] upgrade [package] [package...]"
+    )
+    op.add_option("-v", "--verbose", action="store_true", dest="verbose", default=False)
+    op.add_option("-n", "--dry-run", action="store_true", dest="dry_run", default=False)
+    options, args = op.parse_args()
+
+    if len(args) == 0:
+        op.print_help()
+        sys.exit(1)
+    cmd = args.pop(0)
+    if cmd not in {"compile", "upgrade"}:
+        op.print_help()
+        sys.exit(1)
+
+    _ensure_pip_tools()
+
+    if cmd == "compile":
+        if len(args) != 0:
+            op.print_help()
+            sys.exit(1)
+        compile_source(verbose=options.verbose, dry_run=options.dry_run)
+    elif cmd == "upgrade":
+        if len(args) == 0:
+            resp = input("Are you sure you want to upgrade ALL packages? [y/N] ")
```

Comment:
Why not just forbid this, since `compile` will do the same?

`compile_parser.add_argument('files', nargs='+')` should do the trick.

Reply:
They do different things.

- Plain `compile` will maintain the existing versions unless a newer version is required by a changed (or added) package in the .in files.
- `--upgrade` will upgrade all packages to the latest versions allowed by the constraints from the .in files
- `--upgrade-package` will only upgrade the given packages (which can either be direct or indirect dependencies)

I still have to push some docs how this all is supposed to be used. I'll mention this there.

Reply:
I don't think it makes sense to allow restricting which individual files to compile due to the dependencies. 

If there should be a very special edge case where that would be necessary the person doing that would need to have pretty good knowledge of how pip-compile works internally anyway.